sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp

```diff
@@ -0,0 +1,317 @@
+  switch (Layout) {
+  case matrix_layout::row_major:
+    load_accumulator_layoutT<matrix_layout::row_major>(res, src, stride, sg);
+    break;
+  case matrix_layout::col_major:
+    load_accumulator_layoutT<matrix_layout::col_major>(res, src, stride, sg);
+    break;
+  default:
+    assert(false && "Invalid layout specified!");
+  }
+}
+
+template <
+    typename Group, typename S, typename T, size_t NumRows, size_t NumCols,
+    matrix_use Use, matrix_layout Layout, access::address_space Space,
+    access::decorated IsDecorated,
+    typename = typename std::enable_if_t<(Layout == matrix_layout::row_major ||
+                                          Layout == matrix_layout::col_major)>>
+void load_multiplicand_hip(
+    joint_matrix_hip<S, Use, NumRows, NumCols, Layout> &res,
+    multi_ptr<T, Space, IsDecorated> src, size_t stride, Group &sg) {
+  std::ignore = stride;
+  static_assert(std::is_same_v<S, half> || std::is_same_v<S, bfloat16> ||
+                    std::is_same_v<S, int8_t> || std::is_same_v<S, double>,
+                "Unsupported matrix type!");
+
+  const auto idx = sg.get_group_linear_id() * sg.get_local_range()[0] +
+                   sg.get_local_linear_id();
+
+  if constexpr (std::is_same_v<S, double>) {
+    if constexpr (Layout == matrix_layout::row_major) {
+      res.data[0] = src[idx];
+    } else if constexpr (Layout == matrix_layout::col_major) {
+      res.data[0] = src[(idx % NumRows) * 4 + idx / NumRows];
+    }
+  } else {
+    constexpr int Dim = (NumRows == 16) ? 16 : 32;
+    constexpr int K = (NumRows == 16) ? 16 : 8;
+
+    const auto thread_x = idx % Dim;
+    const auto thread_y = idx / Dim;
+
+    if constexpr (Layout == matrix_layout::col_major) {
+      for (int i = 0; i < 4; ++i) {
+        const int c_idx = thread_x * K + i + thread_y * 4;
+        res.data[i] = src[c_idx];
+      }
+    } else if constexpr (Layout == matrix_layout::row_major) {
+      for (int i = 0; i < 4; ++i) {
+        const int r_idx = thread_x + i * NumCols + thread_y * NumCols * 4;
+        res.data[i] = src[r_idx];
+      }
+    }
+  }
+}
+
+template <typename Group, matrix_layout Layout, typename T, size_t NumRows,
+          size_t NumCols, access::address_space Space,
+          access::decorated IsDecorated>
+void store_layoutT(joint_matrix_hip<T, matrix_use::accumulator, NumRows,
+                                    NumCols, matrix_layout::dynamic> &src,
+                   multi_ptr<T, Space, IsDecorated> dst, size_t stride,
+                   Group &sg) {
+  std::ignore = stride;
+  const auto idx = sg.get_group_linear_id() * sg.get_local_range()[0] +
+                   sg.get_local_linear_id();
+
+  if constexpr (std::is_same_v<T, double>) {
+    if constexpr (NumRows == 16 && NumCols == 16) {
+      const auto thread_x = idx % NumCols;
+      const auto thread_y = idx / NumCols;
+      for (int i = 0; i < 4; ++i) {
+        dst[thread_x + i * 4 * NumCols + thread_y * NumCols] = src.data[i];
```

Comment:
I wonder if this loop gets unrolled automatically or if not whether it affects perf?

Reply:
That's fair point. I'd suggest this to investigated it later.